drivers/ethernet/eth_nxp_enet.c

```diff
@@ -0,0 +1,1046 @@
+	 * only report failure if there is no frame to flush,
+	 * which cannot happen in this context.
+	 */
+	status = ENET_ReadFrame(config->base, &data->enet_handle, NULL,
+					0, RING_ID, NULL);
+	__ASSERT_NO_MSG(status == kStatus_Success);
+error:
+	eth_stats_update_errors_rx(get_iface(data, vlan_tag));
+	return -EIO;
+}
+
+static void eth_nxp_enet_rx_thread(void *arg1, void *unused1, void *unused2)
+{
+	const struct device *dev = arg1;
+	const struct nxp_enet_mac_config *config = dev->config;
+	struct nxp_enet_mac_data *data = dev->data;
+
+	while (1) {
+		if (k_sem_take(&data->rx_thread_sem, K_FOREVER) == 0) {
+			while (eth_nxp_enet_rx(dev) == 1) {
+				;
+			}
+			/* enable the IRQ for RX */
+			ENET_EnableInterrupts(config->base,
+			  kENET_RxFrameInterrupt | kENET_RxBufferInterrupt);
+		}
+	}
+}
+
+/*
+ ****************************
+ * PHY management functions *
+ ****************************
+ */
+
+static int nxp_enet_phy_reset_and_configure(const struct device *phy)
+{
+	int ret = 0;
+
+	/* Reset the PHY */
+	ret = phy_write(phy, MII_BMCR, MII_BMCR_RESET);
+	if (ret) {
+		return ret;
+	}
+
+	/* Configure the PHY */
+	ret = phy_configure_link(phy, LINK_HALF_10BASE_T |
+				LINK_FULL_10BASE_T |
+				LINK_HALF_100BASE_T |
+				LINK_FULL_100BASE_T);
+	if (ret) {
+		return ret;
+	}
+
+	return ret;
+}
+
+static void nxp_enet_phy_cb(const struct device *phy,
+				struct phy_link_state *state,
+				void *eth_dev)
+{
+	struct phy_link_state new_state;
+	const struct device *dev = eth_dev;
+	struct nxp_enet_mac_data *data = dev->data;
+
+	if (!state->is_up) {
+		net_eth_carrier_off(data->iface);
+		nxp_enet_phy_reset_and_configure(phy);
+	} else {
+		net_eth_carrier_on(data->iface);
+	}
+}
+
+
+static int nxp_enet_phy_init(const struct device *dev)
+{
+	const struct nxp_enet_mac_config *config = dev->config;
+	int ret = 0;
+
+	ret = nxp_enet_phy_reset_and_configure(config->phy_dev);
+	if (ret) {
+		return ret;
+	}
+
+	ret = phy_link_callback_set(config->phy_dev, nxp_enet_phy_cb, (void *)dev);
+	if (ret) {
+		return ret;
+	}
+
+	return ret;
+}
+
+/*
+ ****************************
+ * Callbacks and interrupts *
+ ****************************
+ */
+
+static void eth_callback(ENET_Type *base, enet_handle_t *handle,
+#if FSL_FEATURE_ENET_QUEUE > 1
+			 uint32_t ringId,
+#endif /* FSL_FEATURE_ENET_QUEUE > 1 */
+			 enet_event_t event, enet_frame_info_t *frameinfo, void *param)
+{
+	const struct device *dev = param;
+	const struct nxp_enet_mac_config *config = dev->config;
+	struct nxp_enet_mac_data *data = dev->data;
+
+	switch (event) {
+	case kENET_RxEvent:
+		k_sem_give(&data->rx_thread_sem);
+		break;
+	case kENET_TxEvent:
+#if defined(CONFIG_PTP_CLOCK_NXP_ENET) && defined(CONFIG_NET_L2_PTP)
+		ts_register_tx_event(dev, frameinfo);
+#endif
+		/* Free the TX buffer. */
+		k_sem_give(&data->tx_buf_sem);
+		break;
+	case kENET_ErrEvent:
+		/* Error event: BABR/BABT/EBERR/LC/RL/UN/PLR.  */
+		break;
```

Comment:
Will `eth_callback` get called in this case when trying to send a frame? If that would be the case, the `tx_buf_sem` will never be released.
Or does the low level driver guarantee to call `eth_callback` with `kENET_TxEvent`?

Reply:
I will look into this further but in this case, this is not new code, I am only moving it to a different file. I think the answer depends on the hardware behavior which I cannot find documentation on. I am hesitant to just give this semaphore because this error interrupt is also used for many RX error situations, and I do not know if the MAC hardware will continue trying to send a TX frame if for example the FIFO is underrun which would cause this interrupt, so the answer of whether to give this semaphore here depends I think on hardware behavior which I am having trouble finding documentation for. If this is how it was in the old driver I think it is okay to leave it as is for now since this PR is not adding this as a new concept, but I will keep looking into this if I have time.